software/runtime/systolic/matmul.h

```diff
@@ -0,0 +1,632 @@
+// Copyright 2022 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Gua Hao Khov, ETH Zurich
+
+/* This library implements a simple systolic architecture emulation
+ * using global code based orchestration
+ */
+
+/* A is an M x N matrix, B is a N x P matrix, and C is a M x P matrix
+ * C = AB
+ * (max dimension is 16-bit)
+ */
+
+#include "alloc.h"
+#include "printf.h"
+#include "systolic/queue_multi.h"
+
+// Dimensions of square systolic array
+// TODO: SQRT ROOT OF NUM_CORES FOR SYSTOLIC SIZE
+#define SYSTOLIC_SIZE 16
+
+// IMPORTANT: DATA_SIZE of queue_multi.h must be set to 4
+
+// Systolic matrix
+typedef struct {
+  int32_t *matrix;
+  uint32_t num_rows;
+  uint32_t num_cols;
+} systolic_matrix_t;
+
+// Cycle counters for benchmark
+uint32_t bm_h_pop_cnt[SYSTOLIC_SIZE][SYSTOLIC_SIZE];
+uint32_t bm_h_push_cnt[SYSTOLIC_SIZE][SYSTOLIC_SIZE];
+uint32_t bm_v_pop_cnt[SYSTOLIC_SIZE][SYSTOLIC_SIZE];
+uint32_t bm_v_push_cnt[SYSTOLIC_SIZE][SYSTOLIC_SIZE];
+
+// Array of queue ptrs in row-major order
+queue_t *queues_vert[SYSTOLIC_SIZE - 1][SYSTOLIC_SIZE];
+queue_t *queues_horz[SYSTOLIC_SIZE][SYSTOLIC_SIZE - 1];
+
+// Define dump functions via CSR writes
+dump(horz_pop, 0x7D1);
+dump(vert_pop, 0x7D2);
+dump(horz_push, 0x7D3);
+dump(vert_push, 0x7D4);
```

Comment:
Those are the performance counters, right? Maybe we should name them accordingly. I thought the assembler was aware of them.